neural_lifetimes/models/nets/embedder.py

```diff
@@ -64,16 +66,31 @@ def build_parameter_dict(self) -> Dict[str, Any]:
         }
 
     def forward(self, x: Dict[str, torch.Tensor]):
+        combined_emb = []
+
         # batch x num_cont_features
         cf = torch.stack([x[f] for f in self.continuous_features], dim=1)
-        cf[cf.isnan()] = 0
+        cf[cf.isnan()] = 0  # TODO do not do if nan is start token
+
+        cf_emb = F.dropout(F.relu(self.c1(cf)), self.drop_rate, self.training)
+        cf_emb = torch.clip(cf_emb, -65000, 65000)
+        assert not cf_emb.isnan().any(), "First Linear for continuous features contains `NaN` values."
 
-        out = F.dropout(F.relu(self.c1(cf)), self.drop_rate, self.training)
         # batch x embed_dim
-        out = F.dropout(F.relu(self.c2(out)), self.drop_rate, self.training)
-        assert not torch.isnan(out.sum())
+        cf_emb = F.dropout(F.relu(self.c2(cf_emb)), self.drop_rate, self.training)
+        assert not cf_emb.isnan().any(), "Second Linear for continuous features contains `NaN` values."
+        combined_emb.append(cf_emb)
 
+        # out = torch.clip(out, -65000, 65000)
         for name in self.discrete_features:
-            out += F.dropout(self.emb[name](x[name]))
+            disc_emb = F.dropout(self.emb[name](x[name]))
+            assert not disc_emb.isnan().any(), f"Embedding for discrete feature '{name}' contains `NaN` values."
+            combined_emb.append(disc_emb)
+
+        combined_emb = torch.stack(combined_emb, dim=-1)
+        out = self.combine(combined_emb).squeeze()
+        assert not out.isnan().any(), "Combined Feature Embeddings contain `NaN` values."
 
+        out = self.layer_norm(out)  # TODO try removing this again once all features are properly normalized
```

Comment:
this helped stabilise with unnormalised features, but it might be worth trying to remove it again later.

Reply:
Shouldn't the normalisation happen at the beginning? If we have different batches with different discrete_value ranges, this normalisation will give the same  `out` normalised values, right?

Reply:
the normalisation of features should happen in the dataloading per feature. This normalises all features together. I think we should remove this.